### ISSUE TITLE
Question. Some MJ-features as native MAS-features?

### ISSUE THREAD

**Issue by Alexustas** [MOARdV/AvionicsSystems#137]
Is there any chance to get some MJ-features as native MAS-features?
What interests me in order of importance:

1. DeltaV calculation
2. More accurate landing prediction and Landing (or Suicide) Burn calculation
3. SASS attitude control (marked with red)
![bandicam 2018-04-09 13-05-57-897](https://user-images.githubusercontent.com/6713671/38496657-c2bafe50-3c17-11e8-886a-a2a9d5008c93.jpg)

**Comment by MOARdV** [comment]
1. It is hard to create a very accurate total deltaV estimate.  Stage dV is a little bit easier.  I think I can make one, but it will not be as accurate as MechJeb, and it will not like strange rocket designs.
2. For atmosphere landing, no.  That is a very complex computation, and even MechJeb does not do very well with it.  I can maybe improve the accuracy of the predictor for non-atmosphere, which will also improve the predictor when close to landing on a planet, as well.
3. Possible.  If I an going to try to make an ascent autopilot, I will need to learn how to hold headings and control roll.

**Comment by MOARdV** [comment]
`fc.StageDeltaV()` is now MAS native.  I probably will not create MAS-native `fc.DeltaV()`, because it is much more complicated.

I also have a MAS Smart A.S.S. prototype.  It uses stock SAS to provide control, but I can set any orientation (heading, pitch, roll) relative to a reference direction, just like the MechJeb SASS controls.  The following are the possible reference directions:

* REF_INERTIAL - inertial frame of reference, to point a particular direction regardless of where you are.
* REF_ORBIT_PROGRADE - Like SAS Prograde when in Orbit mode.
* REF_ORBIT_HORIZONTAL - Heading is aligned with orbit prograde, but pitch is perpendicular to local up.
* REF_SURFACE_PROGRADE - Like SAS Prograde when in Surface mode.
* REF_SURFACE_HORIZONTAL - Heading is aligned with sufrace prograde, but pitch is perpendicular to local up.
* REF_SURFACE_NORTH - Heading is local North, pitch is horizontal.
* REF_TARGET - Like SAS Target +
* REF_TARGET_RELATIVE_VEL - Like SAS Prograde when in Target mode.
* REF_TARGET_ORIENTATION - Like MechJeb SASS Parallel +
* REF_MANEUVER_NODE - Like SAS Maneuver Node.
* REF_SUN - Points towards the sun.

**Comment by MOARdV** [comment]
MAS Attitude Control autopilot is now on [DropBox](https://www.dropbox.com/s/9cc0g2hyage21cg/AvionicsSystems-0.15.0b1.zip?dl=0).  [This section](https://github.com/MOARdV/AvionicsSystems/wiki/MASFlightComputerProxy#autopilot-category) lists available functions.

**Comment by MOARdV** [comment]
[DropBox](https://www.dropbox.com/s/251iyjvdsdimnvr/AvionicsSystems-0.16.0b1.zip?dl=0) now has a functioning MAS-native maneuver autopilot.

**Comment by Alexustas** [comment]
> MAS-native maneuver autopilot.

where can I read the description of this autopilot? how does it work and what does it actually do?

**Comment by MOARdV** [comment]
> where can I read the description of this autopilot? how does it work and what does it actually do?

Not much description.  `fc.ToggleManeuverPilot()`, `fc.GetManeuverPilotActive()` (part of [Autopilot[(https://github.com/MOARdV/AvionicsSystems/wiki/MASFlightComputerProxy#autopilot-category)) are similar to `mechjeb.ToggleManeuverNodeExecutor()` and `mechjeb.ManeuverNodeExecutorActive()`.

It points the vessel at the maneuver node, and it starts the burn so that the maneuver burn is about 50% complete at T+0.  If you disable SAS, or change SAS modes, or delete the maneuver node, it will cancel the autopilot.

It is still work-in-progress (and it will generate a lot of messages in the log while it is running), and I expect it will need some more tuning to make it better.  I will add a "Warp to Maneuver" command later.

**Comment by MOARdV** [comment]
Going to close this.  Here is the status:

1. Stage DV calculation `fc.DeltaVStage()` in MAS does not require MechJeb.  Total DV is too complicated because of staging, and I do not want to spend a lot of time trying to make it work.  `fc.DeltaV()` will use MechJeb if MechJeb is installed, or it returns 0.  I could maybe add Kerbal Engineer support so that MAS can use either MechJeb or Kerbal Engineer for that calculation.
2. I have made changes to landing prediction that should help accuracy.  It is still not perfect.  Also added `fc.SuicideBurnTime()`.
3. Added arbitrary heading / pitch / roll using many reference directions, like MechJeb.  Uses stock SAS to control the heading.
